ebos/damariswriter.hh

```diff
@@ -0,0 +1,384 @@
+ */
+ 
+ 
+template <class TypeTag>
+class DamarisWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::Grid>,
+                                          GetPropType<TypeTag, Properties::EquilGrid>,
+                                          GetPropType<TypeTag, Properties::GridView>,
+                                          GetPropType<TypeTag, Properties::ElementMapper>,
+                                          GetPropType<TypeTag, Properties::Scalar>>
+{
+    using Simulator = GetPropType<TypeTag, Properties::Simulator>;
+    using GridView = GetPropType<TypeTag, Properties::GridView>;
+    using Grid = GetPropType<TypeTag, Properties::Grid>;
+    using EquilGrid = GetPropType<TypeTag, Properties::EquilGrid>;
+    using Scalar = GetPropType<TypeTag, Properties::Scalar>;
+    using ElementContext = GetPropType<TypeTag, Properties::ElementContext>;
+    using Element = typename GridView::template Codim<0>::Entity;
+    using ElementMapper = GetPropType<TypeTag, Properties::ElementMapper>;
+    
+    using BaseType = EclGenericWriter<Grid,EquilGrid,GridView,ElementMapper,Scalar>;
+    
+public:
+    static void registerParameters()
+    {
+        EWOMS_REGISTER_PARAM(TypeTag, bool, EnableDamarisOutputCollective,
+                             "Write output via Damaris using parallel HDF5 to get single file per timestep instead of one per Damaris core.");
+    }
+
+    // The Simulator object should preferably have been const - the
+    // only reason that is not the case is due to the SummaryState
+    // object owned deep down by the vanguard.
+    DamarisWriter(Simulator& simulator)
+        : BaseType(simulator.vanguard().schedule(),
+                   simulator.vanguard().eclState(),
+                   simulator.vanguard().summaryConfig(),
+                   simulator.vanguard().grid(),
+                   simulator.vanguard().grid().comm().rank() == 0 ? &simulator.vanguard().equilGrid() : nullptr,
+                   simulator.vanguard().gridView(),
+                   simulator.vanguard().cartesianIndexMapper(),
+                   simulator.vanguard().grid().comm().rank() == 0 ? &simulator.vanguard().equilCartesianIndexMapper() : nullptr,
+                   false, false)
+        , simulator_(simulator)
+    {
+        this->damarisUpdate_ = true ;
+        
+        rank_ = simulator_.vanguard().grid().comm().rank() ;
+        nranks_ = simulator_.vanguard().grid().comm().size();
+        
+        const auto& gridView = simulator_.gridView();
+        const auto& interior_elements = elements(gridView, Dune::Partitions::interior);
+        // Get the size of the unique vector elements (excludes the shared 'ghost' elements)
+        numElements_ = std::distance(interior_elements.begin(), interior_elements.end());
+        
+        this->elements_rank_offsets_.resize(nranks_) ;
+        this->damarisOutputModule_ = std::make_unique<EclOutputBlackOilModule<TypeTag>>(simulator, this->collectToIORank_);
```

Comment:
This means that this writer owns a complete set of the output data stored in the EclOutputBlackOilModule (or precisely: in its base class EclGenericOutputBlackoilModule). So that will be duplicated, since also the EclWriter owns one of these.

We should try to avoid this extra copy and the processing required to build it. Maybe the EclProblem class should own it instead, and pass it to the EclWriter and DamarisWriter? Also, the prepareLocalCellData() method is duplicated between the writers, maybe that could be moved to EclProblem instead of duplicating it? (If so, renaming it prepareLocalCellDataForOutput() would make its purpose clear in the new context.)

@bska I would value your input on this. I want to have the separate DamarisWriter, but not to duplicate storage, processing and code more than necessary.

Reply:
Hi, yes, I see that issue.  I thought that this was looked after by passing *data::Solution& localCellData* as a reference from EclProblem, first to DamarisWriter::writeOutput() and then to EclWriter::writeOutput(). However, I think that the allocBuffers() call is where the data is created/copied and it is done by both damarisOutputModule_ and eclOutputModule_.

I'll try again passing through a EclOutputBlackOilModule object from  EclProblem, like you suggest.

Reply:
@bska and I discussed this, and it is clear that we should attempt some refactoring to reduce the complexity of the data flow in this part of the simulator. However, doing so without incurring any extra copying is a bit tricky.

I believe that for now maybe you should not refactor as I suggested above, but instead keep more or less the current design. (We refactor after.) One change we request is that then the call EclWriter::writeOutput() should take a rvalue reference (`&&`) instead of an lvalue reference (`&`) since it eats the data passed into it. Also you must then pass it with std::move(). This makes explicit in the code the comment you already added to EclProblem (keep the comment as well) at the call site.

Reply:
Okay, I can leave it for now. I'm thinking that the DamarisWriter and EclWriter will have to be consolidated into a single class. Not so much of a problem, if it works.